### ISSUE TITLE
Smaller container images

### ISSUE THREAD

**Issue by finkr** [PureStorage-OpenConnect/pure-fa-openmetrics-exporter#25]
Current container images are fairly large (441Mb)  because they contains Go-Lang build stage.

Furthermore the binary (5.54MB) is (almost) statically compiled, so the base image isn't need.

Consider using Docker [multistage](https://docs.docker.com/build/building/multi-stage/#use-multi-stage-builds).

```
$ docker history quay.io/purestorage/pure-fa-om-exporter:latest 
IMAGE          CREATED        CREATED BY                                      SIZE      COMMENT
bdb74f431b02   7 weeks ago    /bin/sh -c #(nop)  CMD ["--host" "0.0.0.0" "…   0B        
<missing>      7 weeks ago    /bin/sh -c #(nop)  ENTRYPOINT ["/usr/local/b…   0B        
<missing>      7 weeks ago    /bin/sh -c #(nop)  EXPOSE 9490                  0B        
<missing>      7 weeks ago    /bin/sh -c go build -v -o /usr/local/bin/pur…   47.7MB    
<missing>      7 weeks ago    /bin/sh -c #(nop) COPY dir:2be547e94338b94e0…   1.28MB    
<missing>      7 weeks ago    /bin/sh -c go mod download && go mod verify     38.9MB    
<missing>      7 weeks ago    /bin/sh -c #(nop) COPY multi:fb0a91644ecf963…   48.9kB    
<missing>      7 weeks ago    /bin/sh -c #(nop) WORKDIR /usr/src/app          0B        
<missing>      8 weeks ago    /bin/sh -c #(nop) WORKDIR /go                   0B        
<missing>      8 weeks ago    /bin/sh -c mkdir -p "$GOPATH/src" "$GOPATH/b…   0B        
<missing>      8 weeks ago    /bin/sh -c #(nop)  ENV PATH=/go/bin:/usr/loc…   0B        
<missing>      8 weeks ago    /bin/sh -c #(nop)  ENV GOPATH=/go               0B        
<missing>      8 weeks ago    /bin/sh -c set -eux;  apk add --no-cache --v…   347MB     
<missing>      8 weeks ago    /bin/sh -c #(nop)  ENV GOLANG_VERSION=1.19.3    0B        
<missing>      8 weeks ago    /bin/sh -c #(nop)  ENV PATH=/usr/local/go/bi…   0B        
<missing>      8 weeks ago    /bin/sh -c set -eux;  if [ -e /etc/nsswitch.…   0B        
<missing>      2 months ago   /bin/sh -c apk add --no-cache ca-certificates   519kB     
<missing>      2 months ago   /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B        
<missing>      2 months ago   /bin/sh -c #(nop) ADD file:ceeb6e8632fafc657…   5.54MB    
```

**Comment by finkr** [comment]
Two possibilities:
* Simple [multistage](https://docs.docker.com/build/building/multi-stage/#use-multi-stage-builds) image, using alpine as base image, includes apk and shell (21Mb)
* Statically compiled with empty base image (scratch), in the spirit of [distroless](https://github.com/GoogleContainerTools/distroless), (14Mb).

### Simple  multistage

```Dockerfile
FROM golang:alpine as build

WORKDIR /usr/src/app

# pre-copy/cache go.mod for pre-downloading dependencies and only redownloading them in subsequent builds if they change
COPY go.mod go.sum ./
RUN go mod download && go mod verify

COPY . .
RUN go build -v -o /usr/local/bin/pure-fa-om-exporter cmd/fa-om-exporter/main.go

FROM alpine
COPY --from=build  /usr/local/bin/pure-fa-om-exporter /usr/local/bin/pure-fa-om-exporter
EXPOSE 9490
ENTRYPOINT ["/usr/local/bin/pure-fa-om-exporter"]
CMD ["--host", "0.0.0.0", "--port", "9490"]
```

```
$ docker history pure-fa-om-exporter 
IMAGE                CREATED BY                                      SIZE      COMMENT
51d99b9251e5   /bin/sh -c #(nop) CMD ["--host" "0.0.0.0" "…   0B        
353a48112424   /bin/sh -c #(nop) ENTRYPOINT ["/usr/local/b…   0B        
d2a24d618855   /bin/sh -c #(nop) EXPOSE 9490                  0B        
1ca44db4f499   /bin/sh -c #(nop) COPY file:30dad3c48b73c1a1…   13.6MB    
042a816809aa   /bin/sh -c #(nop) CMD ["/bin/sh"]              0B        
<missing>      /bin/sh -c #(nop) ADD file:e4d600fc4c9c293ef…   7.05MB   
```

### Statically compiled (14Mb)

Go-lang compilation options for static build are inspired by [prometheus/node-exporter](https://github.com/prometheus/node_exporter/blob/master/.promu-cgo.yml).

Some projects also provides container with a shell (typically busybox)
```Dockerfile
FROM golang:alpine as build

WORKDIR /usr/src/app

# pre-copy/cache go.mod for pre-downloading dependencies and only redownloading them in subsequent builds if they change
COPY go.mod go.sum ./
RUN go mod download && go mod verify

COPY . .
RUN CGO_ENABLED=1 go build -a -tags 'netgo osusergo static_build' -v -o /usr/local/bin/pure-fa-om-exporter cmd/fa-om-exporter/main.go

FROM scratch
#FROM busybox:musl
COPY --from=build  /usr/local/bin/pure-fa-om-exporter /pure-fa-om-exporter
EXPOSE 9490
ENTRYPOINT ["/pure-fa-om-exporter"]
CMD ["--host", "0.0.0.0", "--port", "9490"]
```

```
$ docker history d6274a0aaee5
IMAGE          CREATED BY                                      SIZE      COMMENT
d6274a0aaee5   /bin/sh -c #(nop) CMD ["--host" "0.0.0.0" "…   0B        
7a04a959cb79   /bin/sh -c #(nop) ENTRYPOINT ["/pure-fa-om-…   0B        
3bb9c2a0527f   /bin/sh -c #(nop) EXPOSE 9490                  0B        
03bae96ab57e   /bin/sh -c #(nop) COPY file:4774cc7a4b571ef1…   13.5MB    
```

**Comment by genegr** [comment]
Thanks for noticing this! It is now fixed in v1.0.4.

**Comment by finkr** [comment]
fixed in https://github.com/PureStorage-OpenConnect/pure-fa-openmetrics-exporter/commit/07f77ffbd127f9c88ab82e5bec2d2ef77c3a960b (Thanks to @genegr )
